source/ada/lsp-lal_utils.adb

```diff
@@ -66,4 +70,132 @@ package body LSP.Lal_Utils is
+   -------------------------
+
+   function Find_All_References
+     (Definition         : Defining_Name;
+      Sources            : GNATCOLL.VFS.File_Array_Access;
+      Charset            : String;
+      Include_Definition : Boolean := False)
+         return Ada_Node_Array
+   is
+      Context : constant Analysis_Context := Definition.Unit.Context;
+      Source_Units : Analysis_Unit_Array (Sources'Range);
+   begin
+      for N in Sources'Range loop
+         Source_Units (N) := Context.Get_From_File
+           (Sources (N).Display_Full_Name,
+            Charset => Charset);
+      end loop;
+
+      declare
+         References : constant Ada_Node_Array :=
+           Definition.P_Find_All_References (Source_Units);
+      begin
+         if Include_Definition then
+            return References & (1 => Definition.As_Ada_Node);
+         else
+            return References;
+         end if;
+      end;
+   end Find_All_References;
+
+   -----------------------
+   -- Containing_Entity --
+   -----------------------
+
+   function Containing_Entity (Ref : Ada_Node) return Defining_Name is
+      Parents : constant Ada_Node_Array := Ref.Parents;
+   begin
+      for Parent of Parents loop
+         if Parent.Kind in Ada_Subp_Decl
+                         | Ada_Subp_Body
+                         | Ada_Task_Def
+                         | Ada_Task_Body
+                         | Ada_Package_Body
+                         | Ada_Package_Decl
+         then
+            return Parent.As_Basic_Decl.P_Canonical_Part.P_Defining_Name;
+         end if;
+      end loop;
+
+      return No_Defining_Name;
+   end Containing_Entity;
+
+   ------------------
+   -- Is_Called_By --
+   ------------------
+
+   function Is_Called_By
+     (Name_Node : Name;
+      Sources   : GNATCOLL.VFS.File_Array_Access;
+      Charset   : String)
+      return References_By_Subprogram.Map
+   is
+      use References_By_Subprogram;
+      use References_List;
+      Result     : Map;
+      Definition : Defining_Name;
+   begin
+      if Name_Node = No_Name then
+         return Result;
+      end if;
+
+      --  Attempt to resolve the name, return no results if we can't or if the
+      --  name does not resolve to a subprogram.
+      Definition := Resolve_Name (Name_Node);
+
+      if Definition = No_Defining_Name
+        or else Definition.P_Basic_Decl.Kind not in Ada_Subp_Decl
+      then
+         return Result;
+      end if;
+
+      --  Obtain all the references
+
+      declare
+         Refs : constant Ada_Node_Array := Find_All_References
+           (Definition, Sources, Charset, Include_Definition => False);
+         Containing : Defining_Name;
+      begin
+         --  Go through all references to Name, organising them by containing
+         --  subprogram.
+
+         for Ref of Refs loop
+            --  Only consider references that are calls.
+            --  ??? To be discussed: how do we want to handle an access to
```

Comment:
Well it's not a direct call, and I'm pretty sure it won't be reported by the GNAT Xrefs, but indeed that could be interesting information.

I had a similar thought the other day about calls that go through generic instantiations, such as:

```ada
procedure Test is

   procedure A_Procedure is null;

   generic
      with procedure Foo;
   package A is
      procedure Bar;
   end A;

   package body A is
      procedure Bar is
      begin
         Foo;
      end Bar;
   end A;


   package A_Inst is new A (A_Procedure);

begin
   A_Procedure;
end Test;
```

Here, the call to `Foo` in the instantiation of `A_Inst` is most certainly a call to `A_Procedure`, yet neither GNAT nor LAL do find this at the moment, which is (IMHO) a bug.

Accesses are a more complicated case, because the reference is dynamic, of a similar problem I guess.

Reply:
I agree that we should also provide an answer in your case.

I guess the high-level question is, when an user does an "is_called_by" request, what references are they interested in listing?

My current thinking is that they'd want not see everything - if the subprogram is being passed by access or through generic instantiations, it's better to list simply the point where the 'access is taken or the generics is instantiated than to omit it. Ie, I would consider all references except labels.

Reply:
I think we should go further actually ... :)

For generic instantiations, I'd even go as far as saying that those should be taken into account by find_all_refs (as an option). Not just the actual parameter to the generic instantiation, but like:

> Reference to A_Procedure [as A.Foo through generic instantiation A_Inst]

For pointers, probably find_all_refs shouldn't take them into account, but is_called_by should. For example, in the following example:

```ada
procedure Main is
   type Acc is access procedure;
   procedure Foo is null;
   procedure Bar is null;
   Inst : Acc;
begin
   Inst := Foo'Access;
   Inst.all;
end Main;
```

is_called_by should report something like:

> Possible call to Foo through access to subprogram

The logic would be extremely simple, and would only report a "possible call" for every access to subprograms, reporting the same in the following case:

```ada
procedure Main is
   type Acc is access procedure;
   procedure Foo is null;
   procedure Bar is null;
   Inst : Acc;
begin
   if Some_Condition then
      Inst := Foo'Access;
   else
      Inst := Bar'Access;
   end if;
   Inst.all;
end Main;
```

(this would also report a possible call to `Bar`).

The logic behind this thinking is that when you do "is_called_by", you actually want all places where the subp *might* be called. 

Which also brings us to dispatching subprograms, that are probably not handled by your code correctly right now. Given a base tagged type `Root` with a primitive procedure `Foo`, and two derived types `A` and `B` with overriding implementations for `Foo`, a call to `Root'Class.Foo` should also be taken into account as a possible call to both `A.Foo` and `B.Foo`.

To conclude (and I might/should have started with that), I think this logic should be part of LAL, not the LSP. LAL xref API should have an `is_called_by` function (and I think that's what we decided back in the day during our first meeting, but it wasn't written anywhere). So I think all of that is a job for @Roldak and you should just replace your code with a code to his implem of `is_called_by`.